sklearn/manifold/t_sne.py

```diff
@@ -706,22 +715,43 @@ def _fit(self, X, skip_num_points=0):
         if self.n_iter < 250:
             raise ValueError("n_iter should be at least 250")
 
+        if self.square_distance not in [True, False, 'legacy', 'warn']:
+            raise ValueError("squared_distance should be True, False, 'legacy'"
+                             "or warn")
+
+        if self.square_distance == 'warn':
+            if self.metric != "euclidean":
+                warnings.warn("The default value of square_distance will be"
+                              "set to True in 0.23.", FutureWarning)
+            self.square_distance = True
+
         n_samples = X.shape[0]
 
         neighbors_nn = None
         if self.method == "exact":
             # Retrieve the distance matrix, either using the precomputed one or
             # computing it.
             if self.metric == "precomputed":
-                distances = X
+                if self.square_distance is True:
+                    distances = X**2
+                else:
+                    # self.squared_distance is False or 'legacy'
+                    distances = X
             else:
                 if self.verbose:
                     print("[t-SNE] Computing pairwise distances...")
 
-                if self.metric == "euclidean":
+                if self.square_distance == 'legacy':
+                    if self.metric == "euclidean":
+                        distances = pairwise_distances(X, metric=self.metric,
+                                                       squared=True)
+                    else:
+                        distances = pairwise_distances(X, metric=self.metric)
+                elif self.square_distance is True:
                     distances = pairwise_distances(X, metric=self.metric,
                                                    squared=True)
```

Comment:
I don't think squared is a parameter for most metrics

Reply:
Good catch. My bad didn't check this thoroughly, I have addressed your comments and few other fixes. Will upload it along with the tests by the evening. Thanks!